src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -295,6 +313,88 @@ void ScheduleTask(DateTimeOffset utcNow)
             }
         }
 
+        internal readonly List<WeakReference<Stats>?> _allStats = new();
+        internal readonly Stats _accumulatedStats;
+
+        private (int, int) Sum()
+        {
+            int hits = 0, misses = 0;
+            lock (_allStats)
+            {
+                hits += _accumulatedStats.Hits;
+                misses += _accumulatedStats.Misses;
+                foreach (WeakReference<Stats>? wr in _allStats)
+                {
+                    if (wr is not null && wr.TryGetTarget(out Stats? stats))
+                    {
+                        hits += stats.Hits;
+                        misses += stats.Misses;
+                    }
+                }
+            }
+            return (hits, misses);
+        }
+
+        private ThreadLocal<Stats>? _stats;
+
+        private void Hit()
+        {
+            GetStats().Hits++;
+        }
+
+        private void Miss()
+        {
+            GetStats().Misses++;
+        }
+
+        private Stats GetStats()
+        {
+            if (_stats == null)
+            {
+                return Initialize();
+            }
+            return _stats!.Value!;
+
+            Stats Initialize()
+            {
+                var s = _stats = new ThreadLocal<Stats>(() => new Stats(this));
+                lock (_allStats)
+                {
+                    _allStats.Add(new WeakReference<Stats>(s.Value!));
+                }
+                return s.Value!;
+            }
+        }
+
+        internal sealed class Stats
+        {
+            public int Hits;
+            public int Misses;
+            private MemoryCache _memoryCache;
+            public Stats(MemoryCache memoryCache)
+            {
+                _memoryCache = memoryCache;
+            }
+            ~Stats()
```

Comment:
Not something I would expect to be part of this PR (and maybe not ever) but I noticed the ThreadLocal<T> API has support for a Values property which exposes all the current values of the ThreadLocal across all threads. Unfortunately I don't think you can make use of it because there is no callback when threads exit and their values are removed from that list. If the ThreadLocal API ever supported such a callback in the future then the code here could be simplified to make use of it rather than re-creating the list tracking and cleanup mechanics a 2nd time. I certainly haven't thought through all the implications of adding such a callback so maybe it was already considered in the past and rejected with good reason. @kouvel @stephentoub

Reply:
This looks very similar to what [ThreadInt64PersistentCounter](https://github.com/dotnet/runtime/blob/dd3b61ddc1b1a553efbed8bbdba2938a3dff4140/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadInt64PersistentCounter.cs) is trying to do, to act as a counter for completed thread pool work items. I had tried using `ThreadLocal<T>` initially but it was too slow and on some typical processors even worse than `Interlocked.Increment`. So I had ended up with that. Accessing a thread-local static variable still appears to be much slower in .NET Core than what I would expect, but at least I got similar perf compared to `Interlocked` and it would probably scale a tiny bit better. If `ThreadLocal<T>` is not being used to track values I wonder if perf would be better to just use a thread-local static variable instead.

Reply:
> I wonder if perf would be better to just use a thread-local static variable instead, though I guess that may not be feasible here.

You mean [ThreadStatic] is faster than `ThreadLocal<T>`? Why would it not be feasible? Seems like it would work too.

Reply:
The statistics are being tracked per `MemoryCache`, so they are tracked on the instance rather than as a static variable.